cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionService.java

```diff
@@ -272,32 +304,45 @@ public Transaction getTransaction() {
      *
      * @param succeeded {@code true} if the job execution completed successfully.
      */
-    public void completed(boolean succeeded) {
-      // If this service doesn't own the transaction, then nothing to do. It is for the explicit transaction case.
-      if (!ownTransaction) {
+    public void completed(boolean succeeded) throws TransactionFailureException {
+      // Return if doesn't need to commit the transaction
+      if (!transactionInfo.commitOnJobEnded()) {
         return;
       }
 
-      try {
-        Optional<Transaction> tx = transaction;
-        if (tx == null || !tx.isPresent()) {
-          // No transaction was started for the job, hence nothing to do.
-          return;
-        }
+      // Get the transaction to commit.
+      Optional<Transaction> tx = transaction;
+      if (tx == null || !tx.isPresent()) {
+        // No transaction was started for the job, hence nothing to do.
+        return;
+      }
 
-        Transaction jobTx = tx.get();
+      Transaction jobTx = tx.get();
+      try {
         if (succeeded) {
           LOG.debug("Committing transaction for job {}", jobId);
-          if (!txClient.commit(jobTx)) {
-            // If failed to commit (which it shouldn't since there is no conflict detection), invalidate the tx
+          try {
+            if (!txClient.commit(jobTx)) {
+              // If failed to commit (which it shouldn't since there is no conflict detection), throw exception
+              throw new TransactionFailureException("Failed to commit transaction " + jobTx);
+            }
+            transactionInfo.onTransactionCompleted(succeeded, null);
+          } catch (Throwable t) {
+            // Any failure will invalidate the transaction
             Transactions.invalidateQuietly(txClient, jobTx);
+            // Since the failure is unexpected, propagate it
+            throw t;
           }
         } else {
-          LOG.debug("Aborting transaction for job {}", jobId);
-          txClient.invalidate(jobTx.getWritePointer());
+          LOG.debug("Invalidating transaction for job {}", jobId);
+          if (!txClient.invalidate(jobTx.getWritePointer())) {
+            throw new TransactionFailureException("Failed to invalid traction " + jobTx);
+          }
         }
       } catch (Throwable t) {
-        LOG.error("Failed to {} transaction for job {}", succeeded ? "commit" : "invalidate", jobId, t);
+        TransactionFailureException failureCause = Transactions.asTransactionFailure(t);
```

Comment:
So if it fails during commit, the t will have a message that indicates that we were trying to commit? And the same in other scenarios? It seems that here the old code made an attempt to give some context because t does not provide that.